### PR TITLE
Create finish reset password endpoint

### PR DESCRIPTION
**Pull request by rhiannonstanford**
This PR adds code towards the completion of reset password:
- adds a /finish-reset-password route to App.tsx for the FinishResetPassword component
- creates the frontend component FinishResetPassword.tsx which hits the backend endpoint /validate-token/:token
- creates backend GET endpoint /validate-token/:token which checks to ensure the JWT is valid(exists in the database) and not expired

### CODE REVIEW

**Review by rhiannonstanford — COMMENTED** [review]
Self review.

File: backend/Auth/ResetPassword.ts [hunk 1]
```diff
@@ -9,6 +9,10 @@ import {
 import { sendEmail, baseUrl } from "../Utils/EmailUtils.js";
 import { makeJWT } from "../Utils/JWTUtils.js";
 import { JWTModel } from "../DB/models/JWT";
+import jwt from "jsonwebtoken";
+
+const { verify } = jwt;
+const secret = process.env.JWT_SECRET || "secret"; // JWT_SECRET="JUC2021s3cr3t"
```
Comment (internettrans) [inline backend/Auth/ResetPassword.ts]: I think I'd prefer removing the default value for JWT_SECRET and also the comment that seems to potentially leak a possible value for it. The reason is that this project is open source and so if it ever gets used by people a hacker could look at this code and then think "oh i wonder if this deployed instance has set JWT_SECRET or not. If not, I know what the secret is (`"secret"`). They also would probably see the comment and think "were they silly enough to put their production env variables in a comment?

The best thing to do would probably be to create a `.env.example` file, similar to what leah has done in her not-yet-merged PR at https://github.com/JustUtahCoders/flax-crm/pull/56/files#diff-a3046da0d15a27e89f2afe639b25748a7ad4d9290af3e7b1b6c1a5533c8f0a8c. And that file could be copied by each person to be `.env` in their project, but then in production they'd set it to a different value.

```suggestion
const jwtSecret = process.env.JWT_SECRET
```

File: backend/Auth/ResetPassword.ts [hunk 2]
```diff
@@ -64,3 +68,50 @@ router.post(
     return res.status(204).end();
   }
 );
+
+router.get(
+  "/validate-token/:token",
+  param("token").exists(),
```
Comment (internettrans) [inline backend/Auth/ResetPassword.ts:74]: I think that express itself will guarantee that the `token` param exists, since otherwise express wouldn't match the incoming request to this route. Could could do `param("token").isString().notEmpty()`, but I think that also would be redundant because express params always start as strings and if it were empty it again would not have matched the incoming request to this route.

So I think this is unnecessary, but it also isn't hurting. It's a good practice in general to add validation for all params, query params, and body from the request, so I think leaving this here is fine.

File: backend/Auth/ResetPassword.ts [hunk 3]
```diff
@@ -64,3 +68,50 @@ router.post(
...
+  async (req, res) => {
+    const errors = validationResult(req);
+    if (!errors.isEmpty()) {
+      return invalidRequest(res, errors);
+    }
+    const token = req.params?.token;
```
Comment (internettrans) [inline backend/Auth/ResetPassword.ts]: My understanding is that `req.params` is always defined, so the `?.` is not necessary here. You can even get typescript to be aware of what params and query params are available by providing generics when calling `router.get()`.

If the reason you used the optional chaining operator is to get typescript to be okay, the better solution is just to strictly type the params. See the code below as an example:

https://github.com/JustUtahCoders/flax-crm/blob/9abb1948c270a35475106c445383123bfb2d83b3/backend/Fields/BatchPostFields.ts#L7

File: backend/Utils/EmailUtils.ts [hunk 4]
```diff
@@ -2,7 +2,7 @@ import google from "googleapis";
 import path from "path";
 import { encode } from "js-base64";
 
-export const baseUrl = process.env.SERVER_ORIGIN || "https://localhost:7600";
+export const baseUrl = process.env.SERVER_ORIGIN || "http://localhost:7600";
```
Comment (internettrans) [inline backend/Utils/EmailUtils.ts:5]: Since this environment variable isn't a security concern, I'm fine with having a default value here - a hacker would not benefit from seeing the default value

File: frontend/Auth/FinishResetPassword.tsx [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+  const [finishResetPasswordFormData, setFinishResetPasswordFormData] =
+    useState<FinishResetPasswordFormData>({
+      password: "",
+    });
+
+  const paramString = props.history.location.search;
+  let searchParams = new URLSearchParams(paramString);
+  const token = searchParams.get("jwt") || "MISSING TOKEN";
+
+  const queryFunctionhelper = (queryKey) => {
```
Comment (internettrans) [inline frontend/Auth/FinishResetPassword.tsx]: It's more common just to inline this into where you're calling `useQuery()`. There's nothing wrong with making it a separate variable, though. If you keep it as a separate variable, fix the casing to be camel case with `queryFunctionHelper`

File: frontend/Auth/FinishResetPassword.tsx [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    const token = queryKey["queryKey"][0];
+    return flaxFetch<object>(
+      `/validate-token/${token}?tokenType=passwordReset`,
+      {
+        method: "GET",
+      }
+    );
+  };
+
+  const info = useQuery(token, queryFunctionhelper);
```
Comment (internettrans) [inline frontend/Auth/FinishResetPassword.tsx]: Info isn't a very good variable name because it's so vague. Maybe `tokenQuery`? And then `tokenQuery.data` would be the response data

File: frontend/Auth/FinishResetPassword.tsx [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+      password: "",
+    });
+
+  const paramString = props.history.location.search;
+  let searchParams = new URLSearchParams(paramString);
+  const token = searchParams.get("jwt") || "MISSING TOKEN";
+
+  const queryFunctionhelper = (queryKey) => {
+    const token = queryKey["queryKey"][0];
+    return flaxFetch<object>(
```
Comment (internettrans) [inline frontend/Auth/FinishResetPassword.tsx]: We should make all queries cancelable with Abort Controllers. THis one and the mutation

File: frontend/Auth/FinishResetPassword.tsx [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+    <div className="flex justify-center h-screen p-24 sm:pt-80">
+      <form
+        onSubmit={unary(submitMutation.mutate)}
+        className="relative lg:max-w-sm"
+      >
+        <h1 className="text-gray-500 place-self-start mb-2 text-5xl lg:text-xl">
+          Set New Password
+        </h1>
+        <p className="text-left text-gray-600 py-8 text-4xl lg:text-sm">
+          Enter your new password below.
```
Comment (internettrans) [inline frontend/Auth/FinishResetPassword.tsx]: I wonder if it would be helpful to have an input for the email, but then have it filled in and disabled. That way they'd know which account they're setting a password for. To do that, we'd need to add the `userEmail` to the backend api's response, which would involve getting the user id from the JWT model and then getting the user object for that user id.

I think showing this would be a nice touch, let's add it.

File: frontend/Auth/FinishResetPassword.tsx [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+        token: token,
+      },
+    });
+    return requestPromise;
+  });
+
+  return (
+    <div className="flex justify-center h-screen p-24 sm:pt-80">
+      <form
+        onSubmit={unary(submitMutation.mutate)}
```
Comment (internettrans) [inline frontend/Auth/FinishResetPassword.tsx:67]: We need to redirect them to the home page after the password is set. Is this something that you plan on adding in a future PR? Could be done with a `useEffect` that checks whether the submitMutation succeeded or not, which is probably better than just adding it straight into the useMutation itself

File: frontend/Auth/FinishResetPassword.tsx [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+              setFinishResetPasswordFormData({
+                ...finishResetPasswordFormData,
+                password: evt.target.value,
+              })
+            }
+            required
+          />
+        </FormField>
+
+        <div className="absolute inset-x-0 my-8 bottom-0">
```
Comment (internettrans) [inline frontend/Auth/FinishResetPassword.tsx]: It's a little surprising to me that absolute positioning was needed for the Submit New Password button. Why was this needed?

File: frontend/Auth/FinishResetPassword.tsx [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+        onSubmit={unary(submitMutation.mutate)}
+        className="relative lg:max-w-sm"
+      >
+        <h1 className="text-gray-500 place-self-start mb-2 text-5xl lg:text-xl">
+          Set New Password
+        </h1>
+        <p className="text-left text-gray-600 py-8 text-4xl lg:text-sm">
+          Enter your new password below.
+        </p>
+        <FormField className="mb-40">
```
Comment (internettrans) [inline frontend/Auth/FinishResetPassword.tsx]: This seems to be creating a lot of whitespace under the input, presumably for the absolutely positioned submit button? If so, I think the submit button can be implemented without absolute positioning which would mean this margin bottom can be decreased

File: frontend/Auth/FinishResetPassword.tsx [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+    return requestPromise;
+  });
+
+  return (
+    <div className="flex justify-center h-screen p-24 sm:pt-80">
+      <form
+        onSubmit={unary(submitMutation.mutate)}
+        className="relative lg:max-w-sm"
+      >
+        <h1 className="text-gray-500 place-self-start mb-2 text-5xl lg:text-xl">
```
Comment (internettrans) [inline frontend/Auth/FinishResetPassword.tsx]: I didn't know about `place-self` in css. Just looked it up and will start using it in my code

File: backend/Auth/ResetPassword.ts (hunk 1 above)
Comment (rhiannonstanford) [reply backend/Auth/ResetPassword.ts]: This makes sense!  Thank you!  Leah's example was very helpful.

File: backend/Auth/ResetPassword.ts (hunk 3 above)
Comment (rhiannonstanford) [reply backend/Auth/ResetPassword.ts]: Thank you!  Passing generics for params, fixed the typescript errors.

File: frontend/Auth/FinishResetPassword.tsx (hunk 5 above)
Comment (rhiannonstanford) [reply frontend/Auth/FinishResetPassword.tsx]: Absolutely!  Good catch.  I fixed it in a future PR but will fix it here too.

File: frontend/Auth/FinishResetPassword.tsx (hunk 6 above)
Comment (rhiannonstanford) [reply frontend/Auth/FinishResetPassword.tsx]: Completely agree.  Will change to `tokenValidationResponse`.

File: frontend/Auth/FinishResetPassword.tsx (hunk 9 above)
Comment (rhiannonstanford) [reply frontend/Auth/FinishResetPassword.tsx:67]: Yes, this will be added in a future PR.

File: frontend/Auth/FinishResetPassword.tsx (hunk 11 above)
Comment (rhiannonstanford) [reply frontend/Auth/FinishResetPassword.tsx]: Ah yes, it was was unnecessary.  I reduced it.

File: frontend/Auth/FinishResetPassword.tsx (hunk 10 above)
Comment (rhiannonstanford) [reply frontend/Auth/FinishResetPassword.tsx]: It was needed to ensure that the button stayed at the bottom.  Without it rises to just below the input.
with-absolute
without-absolute

File: frontend/Auth/FinishResetPassword.tsx (hunk 8 above)
Comment (rhiannonstanford) [reply frontend/Auth/FinishResetPassword.tsx]: This is a great idea! I will add it.

File: frontend/Auth/FinishResetPassword.tsx (hunk 7 above)
Comment (rhiannonstanford) [reply frontend/Auth/FinishResetPassword.tsx]: Implemented.

**Review by rhiannonstanford — COMMENTED** [review]
Comments resolved.

File: frontend/Auth/FinishResetPassword.tsx (hunk 10 above)
Comment (internettrans) [reply frontend/Auth/FinishResetPassword.tsx]: I probably would accomplish this with `display: flex` + `flex-direction: column` + `justify-content: space-between`. Using absolute position is fine but it generally complicates things when adding new features or maintaining things. Specifically, having to account for the button when creating padding/margin can be a pain, scroll behaviors are a lot more difficult since absolute positioned elements don't scroll with everything else, and also creating absolute positioned elements relative to the whole body/document is more difficult once the nearest positioned ancestor has changed. So please switch to using the css i described above

Comment (rhiannonstanford) [reply frontend/Auth/FinishResetPassword.tsx]: Removed absolute.  Changed css.  Works great.  Thank you!

File: backend/Auth/ResetPassword.ts [hunk 13]
```diff
@@ -64,3 +68,67 @@ router.post(
...
+router.get<Params>(
+  "/validate-token/:token",
+  param("token").exists(),
+  async (req, res) => {
+    const errors = validationResult(req);
+    if (!errors.isEmpty()) {
+      return invalidRequest(res, errors);
+    }
+    const token = req.params.token;
+    const tokenType = req.query?.tokenType;
```
Comment (internettrans) [inline backend/Auth/ResetPassword.ts]: The optional chaining here could be avoided with better types. On line 72, do `router.get(<Params, void, ResponseBody, QueryParams>)` and then define a typescript interface:

```ts
interface QueryParams {
  tokenType?: string;
}

interface ResponseBody {
  tokenIsValid: boolean;
  tokenIsExpired: boolean;
  email?: string;
}
```

File: backend/Auth/ResetPassword.ts [hunk 14]
```diff
@@ -64,3 +68,67 @@ router.post(
...
+    } else {
+      return res
+        .status(200)
+        .json({ tokenIsValid: false, tokenIsExpired: false });
+    }
+  }
+);
+
+function tokenIsValid(token: string, jwtSecret: string | undefined): boolean {
+  if (jwtSecret === undefined) {
```
Comment (internettrans) [inline backend/Auth/ResetPassword.ts]: It's a little abnormal to specifically check for undefined rather than just truthiness checks. It implies that `null`, `NaN`, and `""` might be valid JWTs and should proceed to the code below.

Does the `verify()` function throw an error when given `undefined`? Or why was this check needed? If we do need this check, let's use a truthiness check unless `null` or other falsey values are possibly valid jwt tokens:

```suggestion
  if (!jwtSecret) {
```

File: frontend/Auth/FinishResetPassword.tsx [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+  const userEmail = tokenValidationResponse?.email;
+
+  const submitMutation = useMutation<
+    void,
+    Error,
+    React.FormEvent<HTMLFormElement>
+  >((evt) => {
+    evt.preventDefault();
+    const ac = new AbortController();
+    let requestPromise = flaxFetch<void>(`/send-reset-password-email`, {
```
Comment (internettrans) [inline frontend/Auth/FinishResetPassword.tsx]: It would maybe be nicer to just skip making the request to the backend if there is no token (rather than making a request with `"MISSING TOKEN"`). Something like this:

```ts
if (token) {
  return flaxFetch(...)
} else {
  return Promise.resolve({
    tokenIsValid: false,
    tokenIsExpired: false,
    email: null
  })
}
```

Comment (internettrans) [reply frontend/Auth/FinishResetPassword.tsx]: Additionally `<void>` is not the correct response body for this request. flaxFetch should allow you to just provide no generic (`flaxFetch()`). If it doesn't, we should fix that.

```suggestion
    let requestPromise = flaxFetch(`/send-reset-password-email`, {
```

**Review by internettrans — COMMENTED** [review]
Small comments

File: backend/Auth/ResetPassword.ts (hunk 14 above)
Comment (rhiannonstanford) [reply backend/Auth/ResetPassword.ts]: I included the check because it resolved type errors on the Line where verify is called.  The errors were: ```    Argument of type 'string | undefined' is not assignable to parameter of type 'Secret | GetPublicKeyOrSecret'.
      Type 'undefined' is not assignable to type 'Secret | GetPublicKeyOrSecret'.``` Thank you! Adjusted to a truthiness check.

File: frontend/Auth/FinishResetPassword.tsx (hunk 15 above)
Comment (rhiannonstanford) [reply frontend/Auth/FinishResetPassword.tsx]: Pair programmed to resolve.

**Review by rhiannonstanford — COMMENTED** [review]
Comments resolved.

**Review by internettrans — APPROVED** [review]
